src/jit/lsra.cpp

```diff
@@ -6644,6 +6657,36 @@ bool LinearScan::isSecondHalfReg(RegRecord* regRec, Interval* interval)
 
     return false;
 }
+
+//--------------------------------------------------------------------------------------
+// canRestorePreviousInterval: Test if we can restore previous interval
+//
+// Arguments:
+//    regRec - a register which contains previous interval to be restored
+//    assignedInterval - an interval just unassigned
+//
+// Assumptions:
+//    None
+//
+// Return Value:
+//    True only if previous interval of regRec can be restored
+//
+bool LinearScan::canRestorePreviousInterval(RegRecord* regRec, Interval* assignedInterval)
+{
+    bool retVal =
+        (regRec->previousInterval != nullptr && regRec->previousInterval != assignedInterval &&
+         regRec->previousInterval->assignedReg == regRec && regRec->previousInterval->getNextRefPosition() != nullptr);
+
+    if (retVal && regRec->previousInterval->registerType == TYP_DOUBLE)
+    {
+        regNumber  nextRegNum = REG_NEXT(regRec->regNum);
+        RegRecord* nextRegRec = getRegisterRecord(nextRegNum);
+
+        retVal = retVal &&
+                 (nextRegRec->assignedInterval == nullptr && regRec->previousInterval == nextRegRec->previousInterval);
```

Comment:
Why is the second half of this conjunction needed? If the register is unassigned (the first condition) then it is safe to reassign it to this `Interval`, even if by chance it has been assigned and unassigned to some other `Interval` (thereby potentially overwriting the `previousInterval` value). Note that the `previousInterval` handling is purely a heuristic to maintain association between an `Interval` and its preferred register.